src/com/maddyhome/idea/vim/KeyHandler.java

```diff
@@ -352,188 +336,261 @@ private void handleEditorReset(@NotNull Editor editor, @NotNull KeyStroke key, @
-          final boolean recursive = mappingInfo.isRecursive() && !(first && fromIsPrefix);
-          handleKey(editor, keyStroke, currentContext, recursive);
-          first = false;
-        }
+
+    if (mappingInfo == null) {
+      return false;
+    }
+
+    mappingState.clearKeys();
+
+    final EditorDataContext currentContext = new EditorDataContext(editor);
+
+    final List<KeyStroke> toKeys = mappingInfo.getToKeys();
+    final VimExtensionHandler extensionHandler = mappingInfo.getExtensionHandler();
+
+    if (toKeys != null) {
+      final boolean fromIsPrefix = isPrefix(mappingInfo.getFromKeys(), toKeys);
+      boolean first = true;
+      for (KeyStroke keyStroke : toKeys) {
+        final boolean recursive = mappingInfo.isRecursive() && !(first && fromIsPrefix);
+        handleKey(editor, keyStroke, currentContext, recursive);
+        first = false;
       }
-      else if (extensionHandler != null) {
-        // Here is a mapping to some vim handler
-        final CommandProcessor processor = CommandProcessor.getInstance();
-        final boolean isPendingMode = CommandState.getInstance(editor).getMappingMode() == MappingMode.OP_PENDING;
-        Map<Caret, Integer> startOffsets =
-          editor.getCaretModel().getAllCarets().stream().collect(Collectors.toMap(Function.identity(), Caret::getOffset));
-
-        if (extensionHandler.isRepeatable()) {
-          VimRepeater.Extension.INSTANCE.clean();
-        }
-        processor.executeCommand(editor.getProject(), () -> extensionHandler.execute(editor, context),
-                                 "Vim " + extensionHandler.getClass().getSimpleName(), null);
+    }
+    else if (extensionHandler != null) {
+      final CommandProcessor processor = CommandProcessor.getInstance();
 
-        if (extensionHandler.isRepeatable()) {
-          VimRepeater.Extension.INSTANCE.setLastExtensionHandler(extensionHandler);
-          VimRepeater.Extension.INSTANCE.setArgumentCaptured(null);
-          VimRepeater.INSTANCE.setRepeatHandler(true);
-        }
-        if (isPendingMode &&
-            !currentCmd.isEmpty() &&
-            currentCmd.peek().getArgument() == null) {
-          Map<Caret, VimSelection> offsets = new HashMap<>();
-
-          for (Caret caret : editor.getCaretModel().getAllCarets()) {
-            @Nullable Integer startOffset = startOffsets.get(caret);
-            if (caret.hasSelection()) {
-              final VimSelection vimSelection = VimSelection.Companion
-                .create(UserDataManager.getVimSelectionStart(caret), caret.getOffset(),
-                        SelectionType.fromSubMode(CommandStateHelper.getSubMode(editor)), editor);
-              offsets.put(caret, vimSelection);
-              commandState.popState();
+      // Cache isOperatorPending in case the extension changes the mode while moving the caret
+      // See CommonExtensionTest
+      // TODO: Is this legal? Should we assert in this case?
```

Comment:
Note for myself: TODO

Reply:
@citizenmatt sorry, what do you mean to assert?

Reply:
My lack of understanding of what's happening here. Later in this method, we're acting based on what the state _was_, not what it currently is, because the extension has changed the mode. Is it correct that the extension can change mode, and is it correct that we ignore the current mode?

`CommonExtensionTest` is changing the mode, but is that because it's a valid thing to do, or is it because it's a convenient way to test this bit of code?

And if it's not valid to change the mode, then I think we should add an assert that the extension did not change it, so that any extension that is incorrectly written will throw an exception during development.

Reply:
Aha, I see. Yes, this is a correct behavior that the extension can change the mode and IdeaVim should handle this case. E.g. suppose that an extension that provides a text object "between commas". So, for this text: `hello, this <caret>is my, text` the "between commas" object would be `this is my`. Since this is a text object, the user should have an ability to use them along with `d`, `y`, `c`, etc. commands. But here is a problem: vim doesn't really provide an API to create a text object. Thus this extension should just add a visual selection for everything between commas. Vim understands it fine and if a selection appears during the operator pending mode, it just applies the command (for example, `d`) to this selected area. So, from the user point of view, it looks like a text object, even there is no command in vim such as `create a new text object`.
The same thing works with `:normal` command (not supported yet). You can map some keys to `:normal T,vt,` (also, select everything between commas), and it will act as a text object:
```
onoremap i, :normal T,vt,
di,    <-- works like `delete between commas`
```

So yes, vim saves the state before the plugin execution (if it was operator pending mode), and continues the command after the plugin execution regardless of the state after the plugin.

Sorry for the confusion. I'll document it a little bit clearer during the reformation for `:normal` command support.